File: Introduction to Data Structures: Hash Table/5. Conclusion/5. Insert Delete GetRandom O(1)/InsertDeleteGetRandom.py
class RandomizedSet:

    def __init__(self):
        self.hashMap = {}
        self.nums = []

    def insert(self, val: int) -> bool:

        valNotInHashMap = val not in self.hashMap.keys()

        if valNotInHashMap:

            self.hashMap[val] = len(self.nums)

            self.nums.append(val)

        return valNotInHashMap

    def remove(self, val: int) -> bool:

        valInHashMap = val in self.hashMap.keys()

        if valInHashMap:

            # get last element in nums and store it in a variable
            lastElement = self.nums[-1]

            # get index of val
            indexOfVal = self.hashMap[val]

            # replace that element with the last element
            self.nums[indexOfVal] = lastElement

            # remove last element
            self.nums.pop()

            # change mapping of last element to the index of val
            self.hashMap[lastElement] = indexOfVal

            # remove mapping of element in hash map
            self.hashMap.pop(val)

        return valInHashMap

    def getRandom(self) -> int:

        return random.choice(self.nums)

# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: Introduction to Data Structures: Hash Table/5. Conclusion/5. Insert Delete GetRandom O(1)/InsertDeleteGetRandom.java

class RandomizedSet {
    
    private Map<Integer, Integer> hashMap;
    private List<Integer> arrayList;
    
    public RandomizedSet() {
        hashMap = new HashMap<Integer, Integer>();
        arrayList = new ArrayList<>();
    }
    
    public boolean insert(int val) {
        
        if (hashMap.containsKey(val)) {
            return false;
        } else {
            arrayList.add(val);
            hashMap.put(val, arrayList.indexOf(val));
            return true;
        }
    }
    
    public boolean remove(int val) {
        if (hashMap.containsKey(val)) {
            
            // get the index of val in arraylist
            int currIndex = hashMap.get(Integer.valueOf(val));
            
            // get the element at the end of array list
            int lastElement = arrayList.get(arrayList.size() - 1);
            
            // replace the removed element with last element
            arrayList.set(currIndex, lastElement);
            
            // remove element at the end of arrayList
            arrayList.remove(arrayList.size() - 1);
            
            // replace mapping for last element
            hashMap.put(lastElement, currIndex);
            
            // remove value from hashMap
            hashMap.remove(Integer.valueOf(val));
                
            return true;
            
        } else {
            
            return false;
        }
    }
    
    public int getRandom() {
        // generate a random number from 0 - max size of arraylist
        int randomIndex = (int) (Math.random() * (arrayList.size()));
        
        // return the element at that index
        return arrayList.get(randomIndex);
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */